source/extensions/filters/network/common/redis/redis_command_stats.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/timespan.h"
+
+#include "common/stats/symbol_table_impl.h"
+
+#include "extensions/filters/network/common/redis/codec.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Common {
+namespace Redis {
+
+class RedisCommandStats {
+public:
+  RedisCommandStats(Stats::Scope& scope, const std::string& prefix, bool enabled);
+
+  Stats::Counter& counter(const Stats::StatNameVec& stat_names);
+  Stats::Histogram& histogram(const Stats::StatNameVec& stat_names);
+  Stats::CompletableTimespanPtr createCommandTimer(std::string command,
+                                                   Envoy::TimeSource& time_source);
+  Stats::CompletableTimespanPtr createAggregateTimer(Envoy::TimeSource& time_source);
+  std::string getCommandFromRequest(const RespValue& request);
+  void updateStatsTotal(const std::string& command);
+  void updateStats(const bool success, const std::string& command);
+  bool enabled() { return enabled_; }
+
+private:
+  Stats::Scope& scope_;
+  Stats::StatNameSet stat_name_set_;
```

Comment:
Sorry that this didn't become clear to me until now.  But I'm wondering if we should have a policy of never recording stats for unexpected commands. In that case, it would be better to use a StatNamePool rather than StatNameSet. Then you would keep your own string->StatName map which you initialize here. If the command-name is not found in the map you would record 'unknown' or similar.

The benefit of StatNameSet is that it lets you transparently deal with stats from builtin names vs stats from newly discovered names. However if you want a policy of folding together all unknown names into a single 'unknown' you can't do it.

Other than that `StatNameSet::add()` is just a wrapper around `StatNamePool::add()` so most of your usage doesn't change.

WDYT?

Reply:
That sounds reasonable, especially as we essentially whitelist what metrics we will use (and any new commands in future redis versions will be added into these lists).

I've followed the example in `grpc/context_impl` on the map/lookup, without mutex as there will be no dynamic additions.